crates/kira/src/sound/streaming/data.rs

```diff
@@ -325,16 +323,21 @@ impl<Error: Send> StreamingSoundData<Error> {
 	*/
 	#[must_use = "This method consumes self and returns a modified StreamingSoundData, so the return value should be used"]
 	pub fn slice(mut self, region: impl IntoOptionalRegion) -> Self {
-		self.slice = region.into_optional_region().map(|Region { start, end }| {
-			let start = start.into_samples(self.decoder.sample_rate());
-			let end = match end {
-				EndPosition::EndOfAudio => self.decoder.num_frames(),
-				EndPosition::Custom(end) => end.into_samples(self.decoder.sample_rate()),
-			};
-			(start, end)
+		self.slice = region.into_optional_region().map(|region| {
+			region.into_samples(self.decoder.num_frames(), self.decoder.sample_rate())
 		});
 		self
 	}
+
+	/// Returns the sample rate from the decoder.
+	pub fn sample_rate(&self) -> u32 {
+		self.decoder.sample_rate()
+	}
+
+	/// Returns the number of frames from the decoder.
+	pub fn decoder_num_frames(&self) -> usize {
```

Comment:
This conflicts with the mental model I'm intending for slicing. If you slice a sound, it's supposed to be as if you made a new sound with only a portion of the original sound. Getting the number of frames from the decoder gives you the number of frames in the original sound. If you want that, just call `num_frames()` on the original, unsliced sound.

Reply:
Long story short I'd like to avoid storing more bytes when sound data types already contain them.

What I'm doing is an alternate audio engine for Cyberpunk 2077 (documentation can be found [here](https://cyb3rpsych0s1s.github.io/audioware/)). I'm looking to make it as flexible for modders as can be, while consuming the least possible resource (CPU, storage, etc).

Currently modders can define both sound + settings in YAML, but also modify them on-the-fly before play, and in the future I'm gonna support modifying them while playing. They also need to be able to retrieve original and sliced duration + settings duration (e.g. loop_region) and I'd like to avoid having to store original slice + modified slice, if possible.

Reply:
That being said I currently store sound data types in 3 ways:
- in-memory `StaticSoundData` already have their `StaticSoundSettings` applied, but yes it's true that I can `.slice(None)` without requiring ownership so it's fine.
- on-demand sound data:
   - static sound: `PathBuf` + `StaticSoundSettings` so here again I can retrieve the original sound duration from `PathBuf` by simply omitting settings.
   - streaming sound: `PathBuf` + StreamingSoundSettings` so here too I can retrieve the original sound duration in the same fashion.

So yeah, I realize that if it conflicts with your mental model of slicing and you would rather not carry on with the changes, I can drop this PR and still work around it without having to allocate.

But being able to get the `Region` duration would still be a good quality-of-life improvement as far as I'm concerned.

Reply:
> They also need to be able to retrieve original and sliced duration + settings duration (e.g. loop_region)

This is the part that I'm unclear about. Are they playing sounds themselves? If so, shouldn't they already know the duration of the loop region they set?

Reply:
Yes this is basically a tool that lets you play your own sounds in Cyberpunk 2077 with kira.
Modders describe their sounds bank in YAML file, and can then play them anytime, including via scripting.

And yes you're right they're supposed to know their sound duration, but in terms of convenience it's always appreciated to be able to retrieve it dynamically, including with contextual parameters (e.g. V's gender, currently spoken locale, etc).